### ISSUE TITLE
require-return-type must not require undefined type

### ISSUE THREAD

**Issue by gajus** [gajus/eslint-plugin-flowtype#3]

**Comment by migueloller** [comment]
There's another issue where if you specify `excludeArrowFunctions: 'expressionsOnly'` it will not check function expressions.

For example:
```es6
const add = function add(x: number, y: number) {
  return x + y;
};
// doesn't report any errors
```

**Comment by danharper** [comment]
Hm, I've added a test for this, and it does error for "Missing return type annotation.":

![image](https://cloud.githubusercontent.com/assets/510740/17605520/afb7c3f2-6012-11e6-8779-2016bdfe1a11.png)

**Comment by migueloller** [comment]
I had forgotten to add `always` in the options array on my `.eslintrc`. Funny though, because `require-parameter-type` worked without it. Is `always` not on by default for `require-return-type`?

**Comment by migueloller** [comment]
In the documentation it shows that `require-param-type` doesn't have an `always` option but `require-return-type` does. Given that both rules are so similar it's a bit confusing that one has it and the other one doesn't. It would make sense if none did IMO.

Also, I have a few more examples (currently failing) that should not error:
```es6
/* @flow */

/**
 *  'flowtype/require-parameter-type': ['error', 'always', {
 *    excludeArrowFunctions: 'expressionsOnly',
 *  }],
 *
 * error: Missing 'x' parameter type annotation.
 */
[].forEach((x) => {
  console.log(x);
});

/**
 *  'flowtype/require-parameter-type': ['error', {
 *    excludeArrowFunctions: 'expressionsOnly',
 *  }],
 *
 * error: Missing 'x' parameter type annotation.
 */
[].forEach((x) => {
  console.log(x);
});

/**
 *  'flowtype/require-parameter-type': ['error', 'always', {
 *    excludeArrowFunctions: true,
 *  }],
 *
 * error: Missing 'x' parameter type annotation.
 */
[].forEach((x) => {
  console.log(x);
});

/**
 *  'flowtype/require-parameter-type': ['error', {
 *    excludeArrowFunctions: true,
 *  }],
 *
 * This one seems to work!
 */
[].forEach((x) => {
  console.log(x);
});

/**
 *  'flowtype/require-parameter-type': ['error', 'always', {
 *    excludeArrowFunctions: 'expressionsOnly',
 *  }],
 *
 * error: Missing 'x' parameter type annotation.
 */
[].forEach((x) => x);

/**
 *  'flowtype/require-parameter-type': ['error', {
 *    excludeArrowFunctions: 'expressionsOnly',
 *  }],
 *
 * This one seems to work!
 */
[].forEach((x) => x);

/**
 *  'flowtype/require-parameter-type': ['error', 'always', {
 *    excludeArrowFunctions: true,
 *  }],
 *
 * error: Missing 'x' parameter type annotation.
 */
[].forEach((x) => x);

/**
 *  'flowtype/require-parameter-type': ['error', {
 *    excludeArrowFunctions: true,
 *  }],
 *
 * This one seems to work!
 */
[].forEach((x) => x);
```
It does look like `require-param-type` doesn't expect the `always` option but it's behaving differently if the function expression has a body (brackets).

**Comment by danharper** [comment]
Ah yeah the two are inconsistent! Will have to address that somehow. I agree `require-return-type` probably shouldn't have that first arg as it wouldn't make sense to say "never have return types" (not sure if it's some sort of eslint convention or something?)

**...and... I've just realised I _completely_ messed up naming that `expressionsOnly` option. It should be something like `bodylessOnly` or something (or something to signify "only when `{ .. }` isn't used)**

I've commented on each of these:

```js
/* @flow */

/**
 * 'flowtype/require-parameter-type': ['error', 'always', {
 *   excludeArrowFunctions: 'expressionsOnly',
 * }],
 *
 * error: Missing 'x' parameter type annotation.
 */
[].forEach((x) => {
  console.log(x);
});

// ^^ incorrect options

/**
 * 'flowtype/require-parameter-type': ['error', {
 *   excludeArrowFunctions: 'expressionsOnly',
 * }],
 *
 * error: Missing 'x' parameter type annotation.
 */
[].forEach((x) => {
  console.log(x);
});

// ^^ should error as "excludeArrowFunctions" is set to arrow function expressions only (bad naming :()

/**
 * 'flowtype/require-parameter-type': ['error', 'always', {
 *   excludeArrowFunctions: true,
 * }],
 *
 * error: Missing 'x' parameter type annotation.
 */
[].forEach((x) => {
  console.log(x);
});

// ^^ incorrect options

/**
 * 'flowtype/require-parameter-type': ['error', {
 *   excludeArrowFunctions: true,
 * }],
 *
 * This one seems to work!
 */
[].forEach((x) => {
  console.log(x);
});

// ^^ yep, options are passed correctly, and you're ignoring all arrow functions

/**
 * 'flowtype/require-parameter-type': ['error', 'always', {
 *   excludeArrowFunctions: 'expressionsOnly',
 * }],
 *
 * error: Missing 'x' parameter type annotation.
 */
[].forEach((x) => x);

// ^^ incorrect options

/**
 * 'flowtype/require-parameter-type': ['error', {
 *   excludeArrowFunctions: 'expressionsOnly',
 * }],
 *
 * This one seems to work!
 */
[].forEach((x) => x);

// ^^ yep, options are passed correctly, and you're ignoring arrow function expressions (bad naming :()

/**
 * 'flowtype/require-parameter-type': ['error', 'always', {
 *   excludeArrowFunctions: true,
 * }],
 *
 * error: Missing 'x' parameter type annotation.
 */
[].forEach((x) => x);

// ^^ yep, options are passed correctly, getting no error for param, getting error for return type

/**
 * 'flowtype/require-parameter-type': ['error', {
 *   excludeArrowFunctions: true,
 * }],
 *
 * This one seems to work!
 */
[].forEach((x) => x);

// ^^ yep, options are passed correctly, and you're ignoring all arrow functions
```

**Comment by danharper** [comment]
It should be `conciseOnly` (arrow funcs have a [concise or block body](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Function_body)) instead of `expressionsOnly`. Dropped the ball there.

**Comment by migueloller** [comment]
Ah, I see! It was weird because technically all arrow functions are [function expressions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/function). Other than that, functions can be function expressions or function declarations.

It would make sense to have the rule not be required for function *expressions* as well since normally those are checked dynamically by Flow. A few examples include higher order functions and function expressions used as callbacks (see the `forEach` example above).

**Comment by migueloller** [comment]
Here's an example:
```es6
type Operation = (x: number, y: number): number;

function addThunk(): Operation {
  return function add(x, y) {
    return x + y;
  };
}
```
Flow will show 100% coverage for this because the `Operation` type takes care of it. Nevertheless, the rules will require `x`, `y`, and `add` to be annotated even though they don't have to. Now, it's entirely the option of the developer if they want to annotate them or not, but being able to opt-out of function expressions would be great! 😄 

For now, I've had to add `eslint-disable-next-line flowtype/require-return-type, flowtype/require-param-type` everywhere.

Thanks for the prompt replies by the way. 🍺

**Comment by migueloller** [comment]
I guess the way to think about it is that perhaps what should be excluded isn't arrow functions, but function expressions themselves. I think a good guideline for this is the [`require-jsdoc`](http://eslint.org/docs/rules/require-jsdoc) rule. It has various options as to when apply it and there has been a great discussion of how to expand on it [here](https://github.com/eslint/eslint/issues/5867).

**Comment by danharper** [comment]
Yeah I got completely thrown off on naming as in the AST only a concise arrow function seems to have `expression: true`, and I ended up calling them that.. lol

Perhaps:

```js
{
  excludeFunctionExpressions: boolean | 'conciseArrowsOnly'
}
```

Although just checking function expressions could probably silence too many uses where they'd be useful to be documented.

**Comment by migueloller** [comment]
Maybe `unblockedArrow`? Or `blockedOnly`?

[This might be a good refence.](http://eslint.org/docs/rules/prefer-arrow-callback.html)

**Comment by migueloller** [comment]
Also, instead of calling it `excludeFunctionExpressions` maybe call it `excludeArrowFunctions` like it is now?

**Comment by migueloller** [comment]
Regarding the usefulness of just checking function expressions I think that it's ok because it's not a default option. If somebody likes writing all of their functions as function expressions they can decide to leave the default value.

I think mainly it makes sense because you cannot annotate a function declaration without annotating the parameters and return type. A function expression though, you can annotate by simply annotating the variable that holds it.

Here's an example:
```es6
// MUST annotate inline
function add(x: number, y: number): number { return x + y; }
```
as opposed to
```es6
type Operation = (x: number, y: number) => number;

const add: Operation = function(x, y) { return x + y };
```
or (much more uncommon)
```es6
const add: (x: number, y: number) => number = function(x, y) { return x + y; };
```

This is the beauty of Flow, option 2 has almost no types in the actual function body but is type checked the exact same as option 1.

**Comment by danharper** [comment]
Unless I'm mistaken, this original issue was fixed in https://github.com/gajus/eslint-plugin-flowtype/commit/6468bcd3f0bea39f13f19673303d6bbd1b6e70b7.
